tests/backends/launcher.py

```diff
@@ -0,0 +1,177 @@
+    return std, err
+
+
+def create_cmd_lines(cmd_line, **kwargs):
+    keys = sorted(kwargs.keys())
+    values_comb = itertools.product(*(kwargs[k] for k in keys))
+    cmd_lines = []
+    for combi in values_comb:
+        temp_cmd_line = cmd_line[::]
+        for key_idx, v in enumerate(combi):
+            k = keys[key_idx]
+            temp_cmd_line = temp_cmd_line.replace(f"[{k}]", str(v))
+        cmd_lines.append(temp_cmd_line)
+    cmd_lines = list(set(cmd_lines))
+    return cmd_lines
+
+
+def undecorated(o):
+    """Remove all decorators from a function, method or class"""
+    # class decorator
+    if type(o) is type:
+        return o
+
+    try:
+        closure = o.__closure__
+    except AttributeError:
+        return
+
+    if closure:
+        for cell in closure:
+            # avoid infinite recursion
+            if cell.cell_contents is o:
+                continue
+
+            # check if the contents looks like a decorator; in that case
+            # we need to go one level down into the dream, otherwise it
+            # might just be a different closed-over variable, which we
+            # can ignore.
+
+            # Note: this favors supporting decorators defined without
+            # @wraps to the detriment of function/method/class closures
+            if looks_like_a_decorator(cell.cell_contents):
+                undecd = undecorated(cell.cell_contents)
+                if undecd:
+                    return undecd
+        else:
+            return o
+    else:
+        return o
+
+
+def looks_like_a_decorator(a):
+    return (
+        isfunction(a) or ismethod(a) or isclass(a)
+    )
+
+
+class DDPLauncher:
+    """ Class used to run ddp tests
+    """
+    @staticmethod
+    def run_from_cmd_line(cli_args:str = None, func_to_run: Optional[Callable] = None, tmpdir: Optional[str] = None, timeout: int = 20):
+        env = os.environ.copy()
+        env["PL_CURRENT_TEST_MODULE"] = str(func_to_run.__module__)
+        env["PL_CURRENT_TEST_NAME"] = str(func_to_run.__name__)
+        return call_training_script(cli_args, tmpdir, env, timeout=timeout)
+
+    def run(cmd_line, **kwargs):
+        cmd_lines = create_cmd_lines(cmd_line, **kwargs)
+
+        def inner(func):
+            @functools.wraps(func)
+            def func_wrapper(*args, **kwargs):
+                tmpdir = kwargs.get("tmpdir")
+                for cmd_line in cmd_lines:
+                    print(f"Launching {func.__name__} with {cmd_line}")
+                    std, err = DDPLauncher.run_from_cmd_line(cmd_line, func, tmpdir, timeout=20)
+
+                    # Make sure the test run properly
+                    result_path = os.path.join(tmpdir, 'ddp.result')
+                    result = torch.load(result_path)
+                    # verify the file wrote the expected outputs
+                    assert result['status'] == 'complete'
+            return func_wrapper
+        return inner
+
+
+def main(args):
```

Comment:
maybe I miss something, is this main used in the test call? if so pls add some docs...

Reply:
as a user, I would start reading from the top-level
1. `if __name__ == "__main__":`
2. `def main(args):`
3. no other motivation to continue :D